mlflow/gluon.py

```diff
@@ -0,0 +1,184 @@
+    """
+    path = os.path.abspath(path)
+    if os.path.exists(path):
+        raise MlflowException("Path '{}' already exists".format(path))
+    data_subpath = "data"
+    data_path = os.path.join(path, data_subpath)
+    os.makedirs(data_path)
+    # The epoch argument of the export method does not play any role in selecting
+    # a specific epoch's paramaters, and is there only for display purposes.
+    gluon_model.export(os.path.join(data_path, _MODEL_SAVE_PATH))
+    with open(os.path.join(path, "architecture.txt"), "w") as fp:
+        fp.write(str(gluon_model))
+    mlflow_model.save(os.path.join(path, "MLmodel"))
+
+
+@experimental
+def log_model(gluon_model: HybridBlock, artifact_path: str):
+    """
+    Log a Gluon model as an MLflow artifact for the current run.
+
+    :param gluon_model: Gluon model to be saved. Must be already hybridized.
+    :param artifact_path: Run-relative artifact path.
+
+    >>> from mxnet.gluon import Trainer
+    >>> from mxnet.gluon.contrib import estimator
+    >>> from mxnet.gluon.loss import SoftmaxCrossEntropyLoss
+    >>> from mxnet.gluon.nn import HybridSequential
+    >>> from mxnet.metric import Accuracy
+    >>> import mlflow
+    >>> # Build, compile, and train your model
+    >>> net = HybridSequential()
+    >>> with net.name_scope():
+    >>> ...
+    >>> net.hybridize()
+    >>> net.collect_params().initialize()
+    >>> softmax_loss = SoftmaxCrossEntropyLoss()
+    >>> trainer = Trainer(net.collect_params())
+    >>> est = estimator.Estimator(net=net, loss=softmax_loss, metrics=Accuracy(), trainer=trainer)
+    >>> est.fit(train_data=train_data, epochs=100, val_data=validation_data)
+    >>> # Log metrics and log the model
+    >>> with mlflow.start_run() as run:
+    >>>   mlflow.gluon.log_model(net, "model")
+    """
+    Model.log(artifact_path=artifact_path, flavor=mlflow.gluon, gluon_model=gluon_model)
+
+
+@experimental
+def autolog():
+    """
+    Enable automatic logging from Gluon to MLflow.
+    Logs loss and any other metrics specified in the fit
+    function, and optimizer data as parameters. Model checkpoints
+    are logged as artifacts to a 'models' directory.
+    """
+    class __MLflowGluonCallback(TrainEnd, EpochEnd):
+        def __init__(self):
+            self.current_epoch = 0
+
+        def epoch_end(self, estimator, *args, **kwargs):
+            logs = {}
+            for metric in estimator.train_metrics:
+                metric_name, metric_val = metric.get()
+                logs[metric_name] = metric_val
+            for metric in estimator.val_metrics:
+                metric_name, metric_val = metric.get()
+                logs[metric_name] = metric_val
+            try_mlflow_log(mlflow.log_metrics, logs, step=self.current_epoch)
+            self.current_epoch += 1
+
+        def train_end(self, estimator, *args, **kwargs):
+            try_mlflow_log(mlflow.log_param, "num_layers", len(estimator.net))
```

Comment:
As is done by https://github.com/mlflow/mlflow/pull/2055 for Keras autologging, maybe we should log these params at the beginning of training rather than at the end (in case training is interrupted / CTRL+C'd etc)

Reply:
I can probably do that. I'll look into it.